README.md
# README

[![Release](https://img.shields.io/github/v/release/JcDenis/commentNotifications)](https://github.com/JcDenis/commentNotifications/releases)
[![Date](https://img.shields.io/github/release-date/JcDenis/commentNotifications)](https://github.com/JcDenis/commentNotifications/releases)
[![Issues](https://img.shields.io/github/issues/JcDenis/commentNotifications)](https://github.com/JcDenis/commentNotifications/issues)
[![Dotclear](https://img.shields.io/badge/dotclear-v2.26-blue.svg)](https://fr.dotclear.org/download)
[![Dotaddict](https://img.shields.io/badge/dotaddict-official-green.svg)](https://plugins.dotaddict.org/dc2/details/commentNotifications)
[![License](https://img.shields.io/github/license/JcDenis/commentNotifications)](https://github.com/JcDenis/commentNotifications/blob/master/LICENSE)

## WHAT IS COMMENTNOTIFICATIONS ?

_commentNotifications_ is a plugin for the open-source 
web publishing software called Dotclear.

It displays notifications on menu and dashboard when new comments arrive.

## REQUIREMENTS

 _commentNotifications_ requires: 

  * usage permissions on blog
  * Dotclear 2.26
  * Rest service available

## USAGE

First install _commentNotifications_, manualy from a zip package or from 
Dotaddict repository. (See Dotclear's documentation to know how do this)

## LINKS

 * License : [GNU GPL v2](https://www.gnu.org/licenses/old-licenses/lgpl-2.0.html)
 * Source & contribution : [GitHub Page](https://github.com/JcDenis/commentNotifications)
 * Packages & details : [Dotaddict Page](https://plugins.dotaddict.org/dc2/details/commentNotifications)

## NOTE

I (JcDenis) am no longer supported this plugin.
You can use plugins dmLastComments and notifyMe to do the same.
Or you can fork this repository and update this plugin.

## CONTRIBUTORS

 * Tomtom (author)
 * Jean-Christian Denis

 You are welcome to contribute to this code.

CHANGELOG.md
0.5 - 2023.04.10
- require dotclear 2.26
- use namespace

0.4 - 2022.12.03
- Update to dotclear 2.24

0.3.3
- Version from Dotaddict

_admin.js
var notificator={css:{menu:{background:'transparent url(index.php?pf=commentNotifications/img/bubble.png) no-repeat top 20px',height:'24px',width:'24px',overflow:'hidden',position:'absolute','text-align':'center','font-weight':'bold',display:'none'}},cookie:'dc_nb_comments',msg:{},init:function(){if($.cookie(notificator.cookie)==null||notificator.reload_nb_comments=='true')$.cookie(notificator.cookie,notificator.nb_comments);var p={top:$('li a[href="comments.php"]').position().top-20,left:$('li a[href="comments.php"]').position().left+$('li a[href="comments.php"]').width()-25};notificator.css.menu=$.extend(notificator.css.menu,p);$('li a[href="comments.php"]').parent().append('<li class="new" rel="comments"></li>');$('li[rel="comments"]').css(notificator.css.menu);notificator.getNbComments()},getNbComments:function(){$.get('services.php',{f:'getNbComments'},function(data){var rsp=$(data).children('rsp')[0];if(rsp.attributes[0].value=='ok'){var delta=$(rsp).find('comments').text()-$.cookie(notificator.cookie);if(delta>0)notificator.applyChanges(delta)};setTimeout('notificator.getNbComments()',10000)})},applyChanges:function(num){var nb_total=parseInt(notificator.nb_comments)+parseInt(num);if(parseInt(notificator.nb_comments)-parseInt($.cookie(notificator.cookie))>0)nb_total=parseInt(notificator.nb_comments);var legend=$('span a[href="comments.php"]'),img=$('a[href="comments.php"] img'),html="";html+=nb_total+" ";html+=notificator.nb_comments>1?notificator.msg.comments:notificator.msg.comment;html+=" ("+num+" ";html+=(num>1?notificator.msg.recents:notificator.msg.recent)+")";legend.html(html);img.attr("src","index.php?pf=commentNotifications/img/comments-b.png").animate();$('li[rel="comments"]').html(num);$('li[rel="comments"]').fadeIn("slow");var r=new RegExp('^\\([0-9]+\\)','g');if($('title').text().match(r)!=null){$('title').text($('title').text().replace(r,'('+num+')'))}else $('title').text('('+num+') '+$('title').text())}};$(document).ready(function(){notificator.init()})

src/Backend.php
<?php
/**
 * @brief commentNotifications, a plugin for Dotclear 2
 *
 * @package Dotclear
 * @subpackage Plugin
 *
 * @author Tomtom and Contributors
 *
 * @copyright Jean-Christian Denis
 * @copyright GPL-2.0 https://www.gnu.org/licenses/gpl-2.0.html
 */
declare(strict_types=1);

namespace Dotclear\Plugin\commentNotifications;

use dcCore;
use dcNsProcess;
use dcPage;

class Backend extends dcNsProcess
{
    public static function init(): bool
    {
        static::$init = defined('DC_CONTEXT_ADMIN')
            && dcCore::app()->auth->check(dcCore::app()->auth->makePermissions([
                dcCore::app()->auth::PERMISSION_USAGE,
            ]), dcCore::app()->blog->id);

        return static::$init;
    }

    public static function process(): bool
    {
        if (!static::$init) {
            return false;
        }

        dcCore::app()->addBehavior('adminPageHTMLHead', function (): void {
            echo
            dcPage::jsModuleLoad(My::id() . '/js/backend.js') .
            '<script type="text/javascript">' . "\n" .
            "//<![CDATA[\n" .
            dcPage::jsVar('notificator.nb_comments', dcCore::app()->blog->getComments([], true)->f(0)) .
            dcPage::jsVar('notificator.reload_nb_comments', (preg_match('/' . preg_quote(dcCore::app()->adminurl->get('admin.comments')) . '(.*)?/', $_SERVER['REQUEST_URI']) ? 'true' : 'false')) .
            dcPage::jsVar('notificator.msg.comment', __('comment')) .
            dcPage::jsVar('notificator.msg.comments', __('comments')) .
            dcPage::jsVar('notificator.msg.recent', __('new')) .
            dcPage::jsVar('notificator.msg.recents', __('news')) .
            "\n//]]>\n" .
            "</script>\n";
        });

        return true;
    }
}


src/Prepend.php
<?php
/**
 * @brief commentNotifications, a plugin for Dotclear 2
 *
 * @package Dotclear
 * @subpackage Plugin
 *
 * @author Tomtom and Contributors
 *
 * @copyright Jean-Christian Denis
 * @copyright GPL-2.0 https://www.gnu.org/licenses/gpl-2.0.html
 */
declare(strict_types=1);

namespace Dotclear\Plugin\commentNotifications;

use dcCore;
use dcNsProcess;
use Dotclear\Helper\Html\XmlTag;

class Prepend extends dcNsProcess
{
    public static function init(): bool
    {
        static::$init = defined('DC_RC_PATH');

        return static::$init;
    }

    public static function process(): bool
    {
        if (!static::$init) {
            return false;
        }

        dcCore::app()->rest->addFunction('getNbComments', function (): XmlTag {
            $rsp = new XmlTag();
            $rsp->comments(dcCore::app()->blog->getComments(['no_content' => true], true)->f(0));

            return $rsp;
        });

        return true;
    }
}


_define.php
<?php
/**
 * @brief commentNotifications, a plugin for Dotclear 2
 *
 * @package Dotclear
 * @subpackage Plugin
 *
 * @author Tomtom and Contributors
 *
 * @copyright Jean-Christian Denis
 * @copyright GPL-2.0 https://www.gnu.org/licenses/gpl-2.0.html
 */
if (!defined('DC_RC_PATH')) {
    return null;
}

$this->registerModule(
    'commentNotifications',
    'Displays notifications on menu and dashboard when new comments arrive',
    'Tomtom and Contributors',
    '0.5-dev',
    [
        'requires'    => [['core', '2.26']],
        'permissions' => dcCore::app()->auth->makePermissions([
            dcAuth::PERMISSION_USAGE,
        ]),
        'type'       => 'plugin',
        'support'    => 'https://github.com/JcDenis/commentNotifications',
        'details'    => 'https://plugins.dotaddict.org/dc2/details/commentNotifications',
        'repository' => 'https://raw.githubusercontent.com/JcDenis/commentNotifications/master/',
    ]
);
